### PR TITLE
[EP] add support for ETP=1

### PR DESCRIPTION
**Pull request by tianyu-l**
This is a followup of original EP support https://github.com/pytorch/torchtitan/pull/1324

### PR summary
[TBA] description + figure

### numerics verification
setup
- optimizer Adam
- steps 100, warmup_steps 20
- seed 42

comparison set
- FSDP 2
- FSDP 2, CP 2, TP 2, EP 8, ETP 1
- FSDP 2 (EP 2), PP 2, TP 2 (ETP 2)

image

### CODE REVIEW

File: torchtitan/config/job_config.py [hunk 1]
```diff
@@ -374,9 +374,20 @@ class Parallelism:
 
     expert_parallel_degree: int = 1
     """
-    Expert parallelism degree. 1 means disabled.
-    Currently, only "dp2ep" is supported, with the following constraints:
-    context_parallel_degree <= expert_parallel_degree <= data_parallel_shard_degree * context_parallel_degree
+    Expert parallelism degree. 1 means disabled. No effect for non-MoE models.
+    Currently, it is supported with the following constraints:
+    - when etp = tp: cp * tp <= ep <= dp_shard * cp * tp
+    - when etp = 1: cp * tp <= ep <= dp_shard * cp * tp
```
Comment (sanketpurandare) [inline torchtitan/config/job_config.py]: Can we also add some comments about the divisibility constraints? For instance, we require ep % cp = 0 and dp_shard * cp % ep == 0. Or do you think since people usually use powers of 2 this is not needed?

Comment (tianyu-l) [reply torchtitan/config/job_config.py]: Sounds good, I can add something. Fwiw I used to use `|` symbol to denote `mod x == 0` but people don't seem to understand what it means.

File: torchtitan/models/moe.py [hunk 2]
```diff
@@ -143,16 +143,17 @@ def init_weights(self, init_std: float):
         nn.init.trunc_normal_(self.w3, mean=0.0, std=init_std)
 
 
-class TokenChoiceTopKRouter(nn.Module):
+class TokenRouter(nn.Module):
```
Comment (sanketpurandare) [inline torchtitan/models/moe.py]: I feel the earlier name was more descriptive, since we are doing the torch.topk operation in this module. TokenRouter seems more like just the routing part of token choice.

File: torchtitan/config/job_config.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/config/job_config.py]: Yeah, I think the `%` symbol is more common?

**Comment by sanketpurandare** [comment]
I added a couple of nit comments. The loss curves look consistent.  LGTM!

**Comment by vwxyzjn** [comment]
@tianyu-l this is very cool. Could you help clarify what's the difference between ETP and EP + TP on a high-level? Do you have an example config to run it, please?
